### PR TITLE
Rework resolver

### LINKED ISSUE #476: [bug] Split doesn't work with main function
It seems that adding a main function makes `String.split` not work.

Here is the working example: https://webassembly.studio/?f=mn7zhulmmq9

If you build and run`testSplit` it returns 13 to the console.
```
export function testSplit(): i32 {
  let str = "hell o o o o - - - - - - - -";
  let strs = str.split(" ");
  return strs.length;
}
// export function main(): void {

// }
```
If you uncomment the main function, build, and rerun it returns be 1.

### PR DESCRIPTION
**Pull request by dcodeIO**
See: https://github.com/AssemblyScript/assemblyscript/issues/473

The previously weird resolver logic has been replaced with less weird tree-based lookups, the internal IR changed accordingly and the AST updated to include nested type names that the resolver can now deal with.

- [x] Finish BuiltinSymbols etc.
- [x] Integrate types into IR to address #144
- [x] Manually check all tests before regenerating fixtures
- [x] Implement nested types names to address #127 and #423
- [x] Double check and clean up

Other notes:
* Also adds support for `export *`, fixes #249
* The `--noTreeShaking` compiler flag has been removed and replaced with an explicit lazy compilation mechanism. Reason is that top-level control flow requires that namespace and static class members are compiled directly if these access mutable top-level variables. Otherwise, their values are wrong when compiled after top-level control flow has concluded.
  ```ts
  var a = 1;
  class A {
    static readonly b = a; // should be 1, but is 2 when compiled lazily
  }
  a = 2;
  ```
  There are ways to detect this condition, of course, but the standard library especially requires an explicit mechanism so we can keep user binaries as bloat-free as possible anyway, so `@lazy` was born. Downside is that compilation of a user program is now more like JS, upside is that compilation is now more like JS. Means: functions are compiled when used only, but globals are compiled sequentially with other top-level logic unless `@lazy`.
* Instead of automagically picking up an exported `main` function by name, and make it call start to execute top-level logic when it's first called, this mechanism has been replaced with a `@start` decorator. Any function can become the function that kicks off the module now and a user who uses `@export function main` for other reasons won't be surprised anymore. fixes #476
* The AST has been slimmed down by removing reverse-linking between individual nodes entirely, which is now exclusively represented by IR. Might make it a little harder to work with, but makes sense for a compiler in order to reduce overall memory footprint significantly respectively to make the AST easily serializable at some point.

### CODE REVIEW

File: src/builtins.ts [hunk 1]
```diff
@@ -3365,19 +3546,8 @@ function deferASM(
...
-  if (names.length > 1) {
-    for (let i = 1; i < names.length; i++) {
-      const subName = names[i];
-      if (prototype && prototype.members) {
-        prototype = assert(prototype.members.get(subName));
-      }
-    }
-  }
+  assert(compiler.program.elementsByName.has(name));
+  var prototype = compiler.program.elementsByName.get(name)!;
```
Comment (MaxGraey) [inline src/builtins.ts]: How about simplify?
```ts
var prototype = assert(compiler.program.elementsByName.has(name));
```

**Comment by dcodeIO** [comment]
Most of the bare features that were necessary are in now. However, this obviously escalated to a point where half the compiler has been rewritten and shit will definitely hit the fan when merged. Going to take some additional time to go through the code step by step next, before regenerating tests and double checking them.

**Comment by dcodeIO** [comment]
Alright, this passes the test suite now. That is, locally and node stable, but not LTS on Travis apparently, which totally makes sense (again). Also, the test suite is terrible. It covers quite some stuff and that's good, better than nothing, but it's unstructured and as such hard to work with and extend. Should either be fewer files with each covering one aspect (it's currently rather randomly distributed) or more folders for specific things, like syntax-level, feature-level, std-level etc.

**Comment by dcodeIO** [comment]
Ok, next up: Nested types for https://github.com/AssemblyScript/assemblyscript/issues/423

**Comment by torch2424** [comment]
WOAH! Export * is now supported!!! 🎉 🎉 🎉 😄 😄 😄 

Made my day, thank you @dcodeIO 👍

**Pull request merged**